src/main/java/com/google/sps/servlets/UserServlet.java

```diff
@@ -90,30 +109,45 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     response.setContentType("application/json; charset=UTF-8");
     response.setCharacterEncoding("UTF-8");
 
-    String json = convertToJson(user);
-    response.getWriter().println(json);
+    response.getWriter().println(user.toJSON());
   }
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-      UserService userService = UserServiceFactory.getUserService();
-      String userId = userService.getCurrentUser().getUserId();
-      String userNickname = request.getParameter(UserInfo.NICKNAME);
+    GoogleIdToken idToken = verifyId(request.getParameter(ID_TOKEN_PARAM));
 
-      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService(); 
+    if (idToken != null) {
+        Payload payload = idToken.getPayload();
+
+        String userId = payload.getSubject();
+        String userNickname = (String) payload.get(NAME);
+        
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService(); 
 
-      Long currentChallengeId = 0L;
-      ArrayList<Integer> challengeStatuses = new ArrayList<Integer>(Collections.nCopies(3, 0));
+        Long currentChallengeId = 0L;
+        ArrayList<Integer> challengeStatuses = new ArrayList<Integer>(Collections.nCopies(3, 0));
 
-      datastore.put(new UserInfo(userId, userNickname, null, null, currentChallengeId, challengeStatuses).toEntity());
+        datastore.put(new UserInfo(userId, userNickname, null, null, currentChallengeId, challengeStatuses, null).toEntity());
+
+    } else {
+        System.out.println("Invalid ID token.");
```

Comment:
nit: (Here and elsewhere) Do you think this output improves debug-ability in the long run? If not, remove.